src/net_processing.cpp

```diff
@@ -2980,8 +2990,18 @@ void ProcessMessage(
         else if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS)
         {
             bool fRejectedParents = false; // It may be the case that the orphans parents have all been rejected
+
+            // Deduplicate parent txids, so that we don't have to loop over
+            // the same parent txid more than once down below.
+            std::vector<uint256> unique_parents;
+            unique_parents.reserve(tx.vin.size());
             for (const CTxIn& txin : tx.vin) {
-                if (recentRejects->contains(txin.prevout.hash)) {
+                unique_parents.push_back(txin.prevout.hash);
```

Comment:
This might be an over-optimization, but I expect that transactions which contain multiple inputs from the same parent may group those inputs together. If that's true, then checking the prevout hash against the last element of `unique_parents` would prevent us from adding duplicates in many cases.

Reply:
concept nack this one: I agree that this might be a good heuristic, but how commonly does a single wallet have multiple payments from the same parent? The need to compare an extra O(N) times would probably be slower than sorting & removing the extra elements anyways. We've already allocated the memory, may as well use it!

If this does come up as a big common case, the approach I'd take is to call std::unique/erase before sorting and after sorting. I suspect this would be faster because all the data is inline at this point and we don't have to contend with branch mispredictions while filling the vector.

But I think it's overkill :)

Reply:
yep, agree